File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type depthNode struct {
    node  *TreeNode
    depth int
}

func rightSideView(root *TreeNode) []int {
    var visibleNodes []int
    
    // === BFS ===
    queue := []depthNode{}
    
    if root != nil {
        queue = append(queue, depthNode{node: root, depth: 0})
    }
    
    for len(queue) > 0 {
        node, depth := queue[0].node, queue[0].depth
        queue = queue[1:] // dequeue
        
        if depth == len(visibleNodes) {
            visibleNodes = append(visibleNodes, node.Val)
        }
        
        if node.Right != nil {
            queue = append(queue, depthNode{node.Right, depth + 1})
        }
        if node.Left != nil {
            queue = append(queue, depthNode{node.Left, depth + 1})
        }
    }
    
    return visibleNodes
}

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
  const visibleNodes: number[] = [];
  
//  === BFS ===
//   const queue = [];
  
//   root && queue.push([root, 0]);
  
//   while (queue.length) {
//     // A real queue would have O(1) shift (dequeue),
//     // but JS Arrays have O(n) shift ¯\_(ツ)_/¯
//     const [node, depth] = queue.shift();
    
//     visibleNodes[depth] = node.val;
    
//     node.left && queue.push([node.left, depth + 1]);
//     node.right && queue.push([node.right, depth + 1]);
//   }
  
//  === DFS ===
  (function dfs(node: TreeNode | null, depth: number) {
    if (!node) {
      return;
    }
    
    visibleNodes[depth] = node.val;
    
    dfs(node.left, depth + 1);
    dfs(node.right, depth + 1);
  })(root, 0);
  
  return visibleNodes;
};